homeassistant/components/goodwe/sensor.py

```diff
@@ -0,0 +1,257 @@
+        self._inverter = inverter
+
+    async def set_work_mode(self, work_mode: int):
+        """Set the inverter work mode."""
+        await self._inverter.set_work_mode(work_mode)
+
+    async def set_ongrid_battery_dod(self, ongrid_battery_dod: int):
+        """Set the on-grid battery dod."""
+        await self._inverter.set_ongrid_battery_dod(ongrid_battery_dod)
+
+    async def set_grid_export_limit(self, grid_export_limit: int):
+        """Set the grid export limit."""
+        await self._inverter.set_grid_export_limit(grid_export_limit)
+
+    @property
+    def available(self):
+        """Return True if entity is available."""
+        return self.coordinator.data is not None
+
+    @property
+    def native_value(self):
+        """Return the value reported by the sensor."""
+        if self.coordinator.data is not None:
+            new_value = self.coordinator.data.get(self._MAIN_ENTITY_SENSOR)
+            # If no new value was provided, keep the previous
+            if new_value is not None:
+                self._attr_native_value = new_value
+
+        return self._attr_native_value
+
+    @property
+    def state_attributes(self):
+        """Return the inverter state attributes."""
+        return self.coordinator.data
+
+    @property
+    def extra_state_attributes(self):
+        """Return the inverter state attributes."""
+        data = {
+            "model": self._inverter.model_name,
+            "serial_number": self._inverter.serial_number,
+            "firmware_version": self._inverter.software_version,
+            "arm_version": self._inverter.arm_version,
+        }
+        return data
+
+
+class InverterSensor(CoordinatorEntity, SensorEntity):
+    """Entity representing individual inverter sensor."""
+
+    def __init__(self, coordinator, device_info, description, inverter):
+        """Initialize an inverter sensor."""
+        super().__init__(coordinator)
+        self.entity_description = description
+        self.entity_id = f".{DOMAIN}_{description.key}"
+        self._attr_unique_id = f"{DOMAIN}-{description.key}-{inverter.serial_number}"
+        self._attr_device_info = device_info
+
+    @property
+    def available(self):
+        """Return True if entity is available."""
+        return self.coordinator.data is not None
+
+    @property
+    def native_value(self):
+        """Return the value reported by the sensor."""
+        if self.coordinator.data is not None:
+            new_value = self.coordinator.data.get(self.entity_description.key)
+            # If no new value was provided, keep the previous
+            if new_value is not None:
+                # Total increasing sensor should never be set to 0
+                if (
+                    self.state_class == STATE_CLASS_TOTAL_INCREASING
+                    and "total" in self.entity_description.key
+                ):
+                    if new_value:
+                        self._attr_native_value = new_value
```

Comment:
`_attr_native_value` won't get used ever. Since the `native_value` method overrides the one in the base class (https://github.com/home-assistant/core/blob/dev/homeassistant/components/sensor/__init__.py#L237-L239), `_attr_native_value` is never used. I'd rework this, starting with a `value = None` at the top and `return value` at the bottom and appropriate setting of `value` in between.

Reply:
Actually, the `_attr_native_value` is (ab)used as a store of previous value.
Because of the native unreliability of UDP and the intentional approach of silently ignoring up to 3 consecutive request failures, we need to keep the previous value. 
So in case of these intermittent UDP failures, the `async_update_data` does not fail, but returns empty dict as result, meaning previous value should be presenter as still valid value. And that one is kept in `_attr_native_value`.

This behavior may sound really odd, but it was fine tuned by experience of different user types, preventing e.g. annoying sensor blinking between real and Unavailable value with aggressive (e.g. 1s) `scan_interval` frequencies, when UDP gets really unreliable and it is acceptable and intentional to show the "stale" value instead of Unavailable for up to 3 consecutive failed updates.

Reply:
IMO you shouldnt abuse the attr_native_value as it generates confusion. Better to create your own variable as gwww indicated.

Reply:
Agree, will change that.

Reply:
Done